app/controllers/uploads_controller.rb

```diff
@@ -109,46 +112,78 @@ def klass
   end
 
   def requirements_messages
-    # this is a call to custom validators that are not listed inside the class
-    custom_batch_validator_messages = "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
     [Upload.valid_col_seps]
-      .push(*validation_messages)
-      .push(*custom_batch_validator_messages)
-      .compact
+      .push(*validation_messages_presence)
+      .push(*validation_messages_numericality)
+      .push(*validation_messages_uniqueness)
+  end
+
+  def batch
+    custom_batch_validator_messages = {
+      message: 'Requirement Description:',
+      value: "#{klass.name}Validator::REQUIREMENT_DESCRIPTIONS".safe_constantize
+    }
+
+    return [custom_batch_validator_messages] unless custom_batch_validator_messages[:value].nil?
+
+    {}
+  end
+
+  def klass_validator(validation_class)
+    klass.validators.map do |validations|
+      affected_attributes(validations) if validation_class == validations.class
+    end.flatten.compact
+  end
+
+  def validation_messages_presence
+    presence = { message: 'These columns must have a value: ', value: [] }
+
+    presence[:value] = klass_validator(ActiveRecord::Validations::PresenceValidator)
+    return [presence] unless presence[:value].empty?
+  end
+
+  def validation_messages_numericality
+    numericality = { message: 'These columns can only contain numeric values: ', value: [] }
+
+    numericality[:value] = klass_validator(ActiveModel::Validations::NumericalityValidator)
+
+    [numericality] unless numericality[:value].empty?
   end
 
-  def validation_messages
+  def validation_messages_uniqueness
+    uniqueness = { message: 'These columns should contain unique values: ', value: [] }
```

Comment:
`'These columns should contain unique values: '` kind of looks like "view" code to me. If we were able to move the `message`'s over to our views, we might be able to do away with the arbitrary `{message:, value:}` data structure altogether which would the objects we're passing to our view a bit more straightforward. It would also make a lot of the other code in this controller that's responsible for creating and manipulating this data structure simpler.